### ISSUE TITLE
Native Rust AD: Proof of Concept (TrackedTensor + backward)

### ISSUE THREAD

**Issue by shinaoka** [shinaoka/ndtensors-rs#54]
## Parent Issue
Part of #52 (AD Roadmap), specifically Phase 3: Native Rust AD.

## Goal

Minimal proof-of-concept implementation of tape-based autodiff in pure Rust.

**Scope**: Simple `TrackedTensor` that can:
1. Track computational graph during forward pass
2. Execute `backward()` to compute gradients
3. Support `contract` operation as first example

**Non-goals for PoC**:
- Checkpointing strategies
- Memory optimization
- Full operation coverage
- Forward-mode AD (separate PoC)

### Core Data Structures

```rust
// crates/ndtensors/src/autodiff/mod.rs

/// Unique identifier for graph nodes
#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug)]
pub struct NodeId(u64);

/// Reference to a node in the computation graph
pub struct NodeRef {
    pub id: NodeId,
    pub order: usize,  // topological order for backward traversal
}

/// Tensor with gradient tracking
pub struct TrackedTensor<T: Scalar> {
    pub(crate) tensor: DenseTensor<T>,
    pub(crate) node: Option<NodeRef>,
    pub(crate) requires_grad: bool,
}

/// Gradient storage (accumulated during backward)
pub struct Gradients {
    grads: HashMap<NodeId, Box<dyn Any + Send>>,
}

/// Trait for backward functions
pub trait GradFn<T: Scalar>: Send + Sync {
    fn backward(&self, grad_output: &DenseTensor<T>, grads: &mut Gradients);
    fn parents(&self) -> Vec<NodeId>;
}
```

### Computation Graph

```rust
/// Global computation graph (thread-local)
pub struct ComputationGraph {
    nodes: HashMap<NodeId, Box<dyn AnyGradFn>>,
    next_id: AtomicU64,
}

thread_local! {
    static GRAPH: RefCell<ComputationGraph> = RefCell::new(ComputationGraph::new());
}
```

### Usage Example

```rust
use ndtensors::autodiff::{TrackedTensor, backward};

// Create tracked tensors
let a = TrackedTensor::from_tensor(tensor_a, true);  // requires_grad=true
let b = TrackedTensor::from_tensor(tensor_b, true);

// Forward pass (builds graph)
let c = tracked_contract(&a, &[1, -1], &b, &[-1, 2])?;

// Backward pass
let grads = backward(&c)?;

// Access gradients
let grad_a = grads.get::<f64>(&a)?;
let grad_b = grads.get::<f64>(&b)?;
```

## Implementation Tasks

- [ ] **Task 1**: Basic types (`NodeId`, `NodeRef`, `TrackedTensor`, `Gradients`)
- [ ] **Task 2**: `ComputationGraph` with thread-local storage
- [ ] **Task 3**: `GradFn` trait and `backward()` function
- [ ] **Task 4**: `tracked_contract` operation using existing `contract_vjp`
- [ ] **Task 5**: Tests (simple contraction, chain of operations)
- [ ] **Task 6**: `no_grad` context for inference

## File Structure

```
crates/ndtensors/src/
├── autodiff/
│   ├── mod.rs          // Re-exports
│   ├── tensor.rs       // TrackedTensor
│   ├── graph.rs        // ComputationGraph, NodeId, NodeRef
│   ├── gradients.rs    // Gradients container
│   ├── backward.rs     // backward() implementation
│   └── ops/
│       ├── mod.rs
│       └── contract.rs // tracked_contract
└── lib.rs              // pub mod autodiff (feature-gated)
```

## Feature Flag

```toml
# Cargo.toml
[features]
default = []
autodiff = []
```

## Key Design Decisions

1. **Simple over optimal**: No checkpointing, no memory optimization for PoC
2. **Reuse existing VJP**: `tracked_contract` calls `contract` + records `contract_vjp` for backward
3. **Thread-local graph**: Avoid Arc/Mutex complexity for PoC
4. **Topological order**: Nodes store `order` for correct backward traversal
5. **Type-erased storage**: `Box<dyn Any>` for gradients to support multiple scalar types

## References

- PyTorch autograd: `extern/pytorch/torch/csrc/autograd/`
- Burn autodiff: `extern/burn/crates/burn-autodiff/`
- Existing `contract_vjp`: `crates/ndtensors/src/contract/naive.rs`
